Add an offset to torch.eye + simplify its implementation

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #85145

Adds an optional `k=0` kwarg to `torch.eye` indicating which diagonal to put the ones on.

PR to land https://github.com/pytorch/pytorch/pull/72443/ as Rohit does not work in the PyTorch project anymore

lezcano committed:
Land "array_api: Add an offset to torch.eye"

PR to land https://github.com/pytorch/pytorch/pull/72443/ as Rohit does not work in the PyTorch project anymore

[ghstack-poisoned]

Lezcano commented:
I think there's some errors. I'll work on fixing them and then we can try to land this one.

lezcano committed:
Update on "Land "array_api: Add an offset to torch.eye""


PR to land https://github.com/pytorch/pytorch/pull/72443/ as Rohit does not work in the PyTorch project anymore

[ghstack-poisoned]

lezcano committed:
Update on "Land "array_api: Add an offset to torch.eye""


PR to land https://github.com/pytorch/pytorch/pull/72443/ as Rohit does not work in the PyTorch project anymore

[ghstack-poisoned]

lezcano committed:
Update on "Land "array_api: Add an offset to torch.eye""


PR to land https://github.com/pytorch/pytorch/pull/72443/ as Rohit does not work in the PyTorch project anymore

[ghstack-poisoned]

mruberry commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -2290,26 +2290,13 @@
     CompositeExplicitAutograd: eye
 
-- func: eye.m(int n, int m, *, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None) -> Tensor
+- func: eye.m_out(int n, int? m=None, *, int k=0, Tensor(a!) out) -> Tensor(a!)
```
Unfortunately I think this will be a BC issue, and we should probably maintain the current ATen signatures + add a note. The internal implementation of the signatures can call a new `eye.m_out` op, if desired.

Lezcano replied:
Ok, I'll revert those changes and mark them with a sad face in `native_functions.yaml`.

Lezcano replied:
After a bit of thinking (and after having added the relevant tests to the OpInfos) I don't think there's a way to implement this function so that it works in both Python an C++, and is also BC. See this comment from Ivan for an example where the previous implementation breaks https://github.com/pytorch/pytorch/pull/72443/files#r924345868.

Lezcano replied:
cc @IvanYashchuk in case he has any ideas.

Lezcano replied:
`LazyTensor` complains as well at https://github.com/pytorch/pytorch/actions/runs/3090061256/jobs/4998701127

Lezcano replied:
Nevermind, I fixed the `LazyTensor` issue. Now, I don't know how to go about the JIT, or whether it can be fixed at all...

lezcano committed:
Update on "Add an offset to torch.eye + simplify its implementation"


PR to land https://github.com/pytorch/pytorch/pull/72443/ as Rohit does not work in the PyTorch project anymore

[ghstack-poisoned]

lezcano committed:
Update on "Add an offset to torch.eye + simplify its implementation"


PR to land https://github.com/pytorch/pytorch/pull/72443/ as Rohit does not work in the PyTorch project anymore

[ghstack-poisoned]

Lezcano commented:
The MPS tests were failing when I dispatched the function to the simple implementation. I left the current implementation (which just supports `k=0`) and left a comment.

lezcano committed:
Update on "Add an offset to torch.eye + simplify its implementation"


PR to land https://github.com/pytorch/pytorch/pull/72443/ as Rohit does not work in the PyTorch project anymore

[ghstack-poisoned]

lezcano committed:
Update on "Add an offset to torch.eye + simplify its implementation"


PR to land https://github.com/pytorch/pytorch/pull/72443/ as Rohit does not work in the PyTorch project anymore

[ghstack-poisoned]

lezcano committed:
Update on "Add an offset to torch.eye + simplify its implementation"


PR to land https://github.com/pytorch/pytorch/pull/72443/ as Rohit does not work in the PyTorch project anymore

[ghstack-poisoned]

lezcano committed:
Update on "Add an offset to torch.eye + simplify its implementation"


PR to land https://github.com/pytorch/pytorch/pull/72443/ as Rohit does not work in the PyTorch project anymore

[ghstack-poisoned]

lezcano commented:
@eellison this PR is stuck on some JIT failures. In particular, [NNC](https://github.com/pytorch/pytorch/actions/runs/3096235185/jobs/5011710134) / [nvFuser](https://github.com/pytorch/pytorch/actions/runs/3096235185/jobs/5012128949) fail because here
https://github.com/pytorch/pytorch/blob/b13b10a8fab83c9c260e16a8cfb4d99140e9352b/torch/jit/_trace.py#L344-L351
even though `func` is bound as `torch.eye(n, k=k)`, with `k` a kwarg, it's still parsed as `torch.eye(n, k)`, with `k` a positional arg.  How could we fix this?
